### PR TITLE
unionArrayBy: Find next 1-bits with countTrailingZeros

### LINKED ISSUE #374: Use `countTrailingZeros` to speed up iterating over bitmap?!
https://github.com/haskell-unordered-containers/unordered-containers/blob/72378268aa6bfb2bbea783c119ca6481918d64fc/Data/HashMap/Internal.hs#L1615-L1643

I don't quite understand this code yet, but the comment reminds me of this use of `countTrailingZeros` to find the next `1` bit:

https://github.com/haskus/packages/blob/1ff5059e66d48144b76523bff1ace81dcd6df634/haskus-binary/src/lib/Haskus/Binary/BitSet.hs#L164-L175

### PR DESCRIPTION
**Pull request by sjakobi**
Closes #374.

### CODE REVIEW

**Comment by sjakobi** [comment]
The benchmark results are pretty surprising: The mean runtime of the `HashMap.union` benchmark has literally doubled from ~60μs to ~120μs!

I think this is due to the weird benchmark data of contiguous `Int`s. Of course it doesn't hurt to increment the index by 1 if the node is full! :/

So, I'll need to improve the benchmark data, maybe simply try with `ByteString`s.

Also, check the generated code.

File: Data/HashMap/Internal.hs [hunk 1]
```diff
@@ -1625,23 +1627,24 @@ unionArrayBy f !b1 !b2 !ary1 !ary2 = A.run $ do
...
-    go 0 0 0 (b' .&. negate b') -- XXX: b' must be non-zero
+    let go !b
+            | b == 0    = return ()
+            | otherwise = do
+                let ba = b1 .&. b2
+                    c  = countTrailingZeros b
+                    m  = bit c
+                    i  = sparseIndex b' m
+                    i1 = sparseIndex b1 m
+                    i2 = sparseIndex b2 m
```
Comment (sjakobi) [inline Data/HashMap/Internal.hs:1638]: No idea whether this is useful or not. Storing and incrementing the 3 indices might be cheap too.

File: Data/HashMap/Internal.hs [hunk 2]
```diff
@@ -1625,23 +1627,24 @@ unionArrayBy f !b1 !b2 !ary1 !ary2 = A.run $ do
...
-            | otherwise     = do
-                A.write mary i =<< A.indexM ary2 i2
-                go (i+1)  i1    (i2+1) (m `unsafeShiftL` 1)
-    go 0 0 0 (b' .&. negate b') -- XXX: b' must be non-zero
+    let go !b
+            | b == 0    = return ()
+            | otherwise = do
+                let ba = b1 .&. b2
+                    c  = countTrailingZeros b
+                    m  = bit c
```
Comment (sjakobi) [inline Data/HashMap/Internal.hs:1635]: I should check whether this involves an unnecessary sign check on `c`.

**Comment by sjakobi** [comment]
> The benchmark results are pretty surprising: The mean runtime of the HashMap.union benchmark has literally doubled from ~60μs to ~120μs!

These numbers are from my desktop with an i7-4790K CPU.

I'm now using my old laptop with an i3-2350M CPU. For the `union.Int` benchmark I'm seeing a slowdown from 128  μs to 277  μs (116%), for the new `union.ByteString` benchmark I'm also seeing a slowdown from 167  μs to 240  μs (43%).

That's pretty terrible. I'll have to look at the generated code. At least I know that it matters!

File: Data/HashMap/Internal.hs (hunk 2 above)
Comment (sjakobi) [reply Data/HashMap/Internal.hs:1635]: It does. :(

```
                       c = word2Int# (ctz# wild) } in
                     case >=# c 0# of {
                       __DEFAULT -> case overflowError of wild1 { };
```

Comment (sjakobi) [reply Data/HashMap/Internal.hs:1635]: I've opened https://gitlab.haskell.org/ghc/ghc/-/issues/21290 about this.

**Comment by sjakobi** [comment]
The latest version (eedc32682e456ec22d5154c4693b0f9058faf4ca) finally brings a speed up for the `union.ByteString` benchmark: 167μs -> 143 μs (-14%).

The `union.Int` benchmark is still slower at 128μs -> 140μs (+9%), but IMHO less realistic.

(Both measurements are from the ancient laptop.)

I should try another version that computes the indices instead of having them as arguments in the loop.

It might also be interesting to combine the various bitmap checks into a single case expression. Maybe GHC can generate something like a lookup table from this?!

**Comment by sjakobi** [comment]
> I should try another version that computes the indices instead of having them as arguments in the loop.

Much worse, even if I keep `i` as a loop variable.

**Comment by sjakobi** [comment]
> The latest version ([eedc326](https://github.com/haskell-unordered-containers/unordered-containers/commit/eedc32682e456ec22d5154c4693b0f9058faf4ca)) finally brings a speed up for the `union.ByteString` benchmark: 167μs -> 143 μs (-14%).
> 
> The `union.Int` benchmark is still slower at 128μs -> 140μs (+9%), but IMHO less realistic.
> 
> (Both measurements are from the ancient laptop.)

On my relatively beefier machine with the i7-4790K I'm getting the following numbers:

* `union.Int`: 60μs -> 66μs (+10%)
* `union.ByteString`: 78μs -> 66μs (-15%)

**Comment by sjakobi** [comment]
> It might also be interesting to combine the various bitmap checks into a single case expression. Maybe GHC can generate something like a lookup table from this?!

I tried this:

```
    let go !i !i1 !i2 !b
            | b == 0 = return ()
            | otherwise = do
                let m = 1 `unsafeShiftL` countTrailingZeros b
                let testBit :: Word -> Word
                    testBit x = fromIntegral (fromEnum (x .&. m /= 0))
                let b'' = b .&. complement m
                let t1 = testBit b1
                let t2 = testBit b2 `unsafeShiftL` 1
                case t1 .|. t2 of
                    1 -> do
                        A.write mary i =<< A.indexM ary1 i1
                        go (i+1) (i1+1)  i2    b''
                    2 -> do
                        A.write mary i =<< A.indexM ary2 i2
                        go (i+1)  i1    (i2+1) b''
                    _ -> do
                        x1 <- A.indexM ary1 i1
                        x2 <- A.indexM ary2 i2
                        A.write mary i $! f x1 x2
                        go (i+1) (i1+1) (i2+1) b''
    go 0 0 0 b'
```

Unfortunately it's slower:

```
    union
      Int:        OK (1.78s)
        162  μs ± 1.6 μs
      ByteString: OK (10.76s)
        156  μs ± 1.5 μs
```

File: Data/HashMap/Internal.hs [hunk 3]
```diff
@@ -1622,26 +1622,27 @@ unionArrayBy :: (a -> a -> a) -> Bitmap -> Bitmap -> A.Array a -> A.Array a
...
-            | otherwise     = do
+                go (i+1) (i1+1) i2 b'
+            | otherwise = do
                 A.write mary i =<< A.indexM ary2 i2
-                go (i+1)  i1    (i2+1) (m `unsafeShiftL` 1)
-    go 0 0 0 (b' .&. negate b') -- XXX: b' must be non-zero
+                go (i+1) i1 (i2+1) b'
+          where
+            m = 1 `unsafeShiftL` (countTrailingZeros b)
+            testBit x = x .&. m /= 0
```
Comment (treeowl) [inline Data/HashMap/Internal.hs:1643]: Is performance the same for

```haskell
testBit x = (x `unsafeShiftR` countTrailingZeros b) .&. 1 /= 0
```

Comment (sjakobi) [reply Data/HashMap/Internal.hs:1643]: It's about 5% slower:

```
    union
      Int:        OK (9.74s)
        72.7 μs ± 269 ns
      ByteString: OK (9.44s)
        70.4 μs ± 250 ns
```

**Comment by sjakobi** [comment]
I think I'll leave it at this for now. Ultimately it would be nice to recover the perf losses for cases like the `union.Int` benchmark, but I'll leave that for future work.

**Pull request merged**